common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java

```diff
@@ -0,0 +1,96 @@
+            long totalDelayMs = Long.parseLong(retryOptions.getOrDefault(TOTAL_DELAY_MS_KEY, DEFAULT_TOTAL_DELAY_MS));
+            this.exponential = new Exponential(baseDelayMs, totalDelayMs);
+        }
+
+        @Override
+        protected boolean shouldRetry() {
+            long pauseTimeMs = exponential.get(attempts);
+            if (pauseTimeMs > 0) {
+                Uninterruptibles.sleepUninterruptibly(pauseTimeMs, TimeUnit.MILLISECONDS);
+                attempts += 1;
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s(baseDelayMs: %s, totalDelayMs: %s, currentAttempts: %s)",
+                                 this.getClass().getSimpleName(), exponential.baseDelayMs, exponential.totalDelayMs, attempts);
+        }
+    }
+
+    /**
+     * Calculate the pause time exponentially, according to the attempts.
+     * The total delay is capped at totalDelayMs, meaning the sum of all the previous pauses cannot exceed it.
+     */
+    static class Exponential {
+        // base delay in ms used to calculate the next pause time
+        private final long baseDelayMs;
+        // total delay in ms permitted
+        private final long totalDelayMs;
+
+        Exponential(long baseDelayMs, long totalDelayMs) {
+            Preconditions.checkArgument(baseDelayMs <= totalDelayMs, "baseDelayMs cannot be greater than totalDelayMs");
+            this.baseDelayMs = baseDelayMs;
+            this.totalDelayMs = totalDelayMs;
+        }
+
+        /**
+         * Calculate the pause time based on attempts.
+         * It is guaranteed that the all the pauses do not exceed totalDelayMs.
+         * @param attempts, number of attempts, starts with 0.
+         * @return the next pasuse time in milliseconds, or negtive if no longer allowed.
+         */
+        long get(int attempts) {
+            long nextMaybe = baseDelayMs * (1L << attempts); // Do not care about overflow. pausedInTotal() corrects the value
+            if (attempts == 0) { // first retry
+                return nextMaybe;
+            } else {
+                long pausedInTotal = pausedInTotal(attempts);
+                if (pausedInTotal < totalDelayMs) {
+                    return Math.min(totalDelayMs - pausedInTotal, nextMaybe); // adjust the next pause time if possible
+                }
+                return -1; // the previous retries have exhausted the permits
+            }
+        }
+
+        // Returns the total pause time according to the `attempts`,
+        // i.e. [0, attempts), which is guaranteed to be greater than or equal to 0.
+        // No overflow can happen.
+        private long pausedInTotal(int attempts) {
+            if (attempts >= Long.SIZE) return totalDelayMs; // take care of overflow. Such long pause time is not realistic though.
```

Comment:
```if (attempts >= Long.numberOfLeadingZeros(baseDelayMs - 1)) return totalDelayMs;``` if we want to make sure no overflow can occur

Reply:
mostly agree. It should have taken `baseDelayMs` into consideration. 
However, I do not think the subtraction is needed. The first attempt to start to return the `total` directly is the one that could make the value wraps around, from a positive to a negative.